src/mw_log_subscriber/src/lib.rs

```diff
@@ -0,0 +1,193 @@
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache License Version 2.0 which is available at
+// <https://www.apache.org/licenses/LICENSE-2.0>
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+mod mw_log_ffi;
+
+use crate::mw_log_ffi::*;
+
+use core::ffi::c_char;
+use core::fmt::{self, Write};
+use mw_log::{Level, Log, Metadata, Record};
+use std::ffi::CString;
+
+const MSG_SIZE: usize = 512;
+
+/// Builder for the MwLogger
+pub struct MwLoggerBuilder {
+    context: Option<CString>,
+}
+
+impl MwLoggerBuilder {
+    pub fn new() -> Self {
+        Self { context: None }
+    }
+
+    /// Builds the MwLogger with the specified context and configuration and returns it.
+    pub fn build<const SHOW_MODULE: bool, const SHOW_FILE: bool, const SHOW_LINE: bool>(
+        &self,
+    ) -> MwLogger {
+        let context_cstr = self
+            .context
+            .as_ref()
+            .cloned()
+            .unwrap_or_else(|| CString::new("DFLT").unwrap());
+        let c_logger_ptr = unsafe { mw_log_create_logger(context_cstr.as_ptr() as *const _) };
+        MwLogger {
+            ptr: c_logger_ptr,
+            log_fn: log::<SHOW_MODULE, SHOW_FILE, SHOW_LINE>,
+        }
+    }
+
+    /// Builds and sets the MwLogger as the default logger with the specified configuration.
+    pub fn set_as_default_logger<
+        const SHOW_MODULE: bool,
+        const SHOW_FILE: bool,
+        const SHOW_LINE: bool,
+    >(
+        self,
+    ) {
+        let logger = self.build::<SHOW_MODULE, SHOW_FILE, SHOW_LINE>();
+        mw_log::set_max_level(mw_log_logger_level(logger.ptr));
+        mw_log::set_boxed_logger(Box::new(logger))
+            .expect("Failed to initialize MwLogger as default logger - logger may already be set");
+    }
+
+    /// Sets the context for currently build logger.
+    pub fn context(mut self, context: &str) -> Self {
+        self.context = Some(CString::new(context).expect(
+            "Failed to create CString:
+             input contains null bytes",
+        ));
+        self
+    }
+}
+
+struct BufWriter<const BUF_SIZE: usize> {
+    buf: [u8; BUF_SIZE],
+    pos: usize,
+}
+
+impl<const BUF_SIZE: usize> BufWriter<BUF_SIZE> {
+    fn new() -> Self {
+        Self {
+            buf: [0; BUF_SIZE],
+            pos: 0,
+        }
+    }
+
+    /// Returns a pointer to a null-terminated C string view of the buffer.
+    fn as_c_str_ptr(&self) -> *const c_char {
+        self.buf.as_ptr() as *const c_char
+    }
+
+    fn len(&self) -> usize {
+        self.pos
+    }
+
+    fn revert_pos(&mut self, cnt: usize) {
+        self.pos = self.pos.saturating_sub(cnt);
+    }
+}
+
+impl<const BUF_SIZE: usize> Write for BufWriter<BUF_SIZE> {
+    fn write_str(&mut self, s: &str) -> fmt::Result {
+        let bytes = s.as_bytes();
+        let to_write = bytes.len().min(self.buf.len() - self.pos - 2); // Cutting write, instead error when we don't fit
```

Comment:
I think this should be `- 1` instead of `- 2`. For example, if the buffer is **10** bytes long and it's at position **2**, then we can write **10 - 2 - 1 == 7** more bytes into it and still have one left for the null-terminator.

```suggestion
        let to_write = bytes.len().min(self.buf.len() - self.pos - 1); // Cutting write, instead error when we don't fit
```